File: visual-dotnet/SauceLabs.Visual/Models/SelectiveRegion.cs
using System;
using System.Collections.Generic;
using OpenQA.Selenium;
using SauceLabs.Visual.GraphQL;
using SauceLabs.Visual.Utils;

namespace SauceLabs.Visual.Models
{
    /// <summary>
    /// SelectiveRegion describe a region where change kind can be filtered.
    /// </summary>
    public class SelectiveRegion
    {
        internal Region? Region { get; }
        internal IWebElement? Element { get; }

        internal DiffingOption? EnableOnly { get; }
        internal DiffingOption? DisableOnly { get; }

        private SelectiveRegion(IWebElement element, DiffingOption? enableOnly, DiffingOption? disableOnly)
        {
            Element = element;
            EnableOnly = enableOnly;
            DisableOnly = disableOnly;
        }

        private SelectiveRegion(Region region, DiffingOption? enableOnly, DiffingOption? disableOnly)
        {
            Region = region;
            EnableOnly = enableOnly;
            DisableOnly = disableOnly;
        }

        public static SelectiveRegion EnabledFor(IWebElement element)
        {
            return new SelectiveRegion(element, null, null);
        }

        public static SelectiveRegion EnabledFor(IWebElement element, DiffingOption flags)
        {
            return new SelectiveRegion(element, flags, null);
        }

        public static SelectiveRegion EnabledFor(Region region)
        {
            return new SelectiveRegion(region, DiffingOption.None, null);
        }

        public static SelectiveRegion EnabledFor(Region region, DiffingOption flags)
        {
            return new SelectiveRegion(region, flags, null);
        }

        public static SelectiveRegion DisabledFor(IWebElement element)
        {
            return new SelectiveRegion(element, null, DiffingOption.None);
        }

        public static SelectiveRegion DisabledFor(IWebElement element, DiffingOption flags)
        {
            return new SelectiveRegion(element, null, flags);
        }

        public static SelectiveRegion DisabledFor(Region region)
        {
            return new SelectiveRegion(region, null, DiffingOption.None);
        }

        public static SelectiveRegion DisabledFor(Region region, DiffingOption flags)
        {
            return new SelectiveRegion(region, null, flags);
        }

        internal RegionIn ToRegionIn()
        {
            var diffingOptions = DiffingOptionsInHelper.CreateFromEnableOnlyDisable(EnableOnly, DisableOnly);
            if (Region != null)
            {
                return new RegionIn(Region, diffingOptions);
            }

            if (Element != null)
            {
                return new RegionIn(Element, diffingOptions);
            }

            throw new VisualClientException("No Region has been passed");
        }


        internal ElementIn ToElementIn()
        {
            var diffingOptions = DiffingOptionsInHelper.CreateFromEnableOnlyDisable(EnableOnly, DisableOnly);

            if (Element != null)
            {
                return new ElementIn(Element, null, diffingOptions);
            }

            throw new VisualClientException("No Element has been passed");
        }
    }
}

File: visual-espresso/visual/src/main/java/com/saucelabs/visual/model/SelectiveRegion.java
package com.saucelabs.visual.model;

import android.view.View;

import com.saucelabs.visual.exception.VisualApiException;
import com.saucelabs.visual.graphql.type.RegionIn;
import com.saucelabs.visual.utils.DiffingOptionsHelper;
import com.saucelabs.visual.utils.RegionInFactory;

import org.hamcrest.Matcher;

import java.util.EnumSet;

public class SelectiveRegion {

    private Matcher<View> viewMatcher;
    private View view;

    private final EnumSet<DiffingOption> enableOnly;
    private final EnumSet<DiffingOption> disableOnly;

    private SelectiveRegion(Matcher<View> viewMatcher, EnumSet<DiffingOption> enableOnly, EnumSet<DiffingOption> disableOnly) {
        this.viewMatcher = viewMatcher;
        this.enableOnly = enableOnly;
        this.disableOnly = disableOnly;
    }

    private SelectiveRegion(View view, EnumSet<DiffingOption> enableOnly, EnumSet<DiffingOption> disableOnly) {
        this.view = view;
        this.enableOnly = enableOnly;
        this.disableOnly = disableOnly;
    }

    public static SelectiveRegion enabledFor(Matcher<View> viewMatcher) {
        return new SelectiveRegion(viewMatcher, EnumSet.noneOf(DiffingOption.class), null);
    }

    public static SelectiveRegion enabledFor(Matcher<View> viewMatcher, DiffingOption... options) {
        return new SelectiveRegion(viewMatcher, EnumSet.of(options[0], options), null);
    }

    public static SelectiveRegion enabledFor(View view) {
        return new SelectiveRegion(view, EnumSet.noneOf(DiffingOption.class), null);
    }

    public static SelectiveRegion enabledFor(View view, DiffingOption... options) {
        return new SelectiveRegion(view, EnumSet.of(options[0], options), null);
    }

    public static SelectiveRegion disabledFor(Matcher<View> viewMatcher) {
        return new SelectiveRegion(viewMatcher, null, EnumSet.noneOf(DiffingOption.class));
    }

    public static SelectiveRegion disabledFor(Matcher<View> viewMatcher, DiffingOption... options) {
        return new SelectiveRegion(viewMatcher, null, EnumSet.of(options[0], options));
    }

    public static SelectiveRegion disabledFor(View view) {
        return new SelectiveRegion(view, null, EnumSet.noneOf(DiffingOption.class));
    }

    public static SelectiveRegion disabledFor(View view, DiffingOption... options) {
        return new SelectiveRegion(view, null, EnumSet.of(options[0], options));
    }

    public RegionIn toRegionIn() {
        if (this.view != null) {
            return RegionInFactory.fromView(view, DiffingOptionsHelper.toDiffingOptionsIn(this.enableOnly, this.disableOnly));
        }
        if (this.viewMatcher != null) {
            return RegionInFactory.fromViewMatcher(viewMatcher, DiffingOptionsHelper.toDiffingOptionsIn(this.enableOnly, this.disableOnly));
        }
        throw new VisualApiException("No region has been provided");
    }

}
